src/inmemory_cache.rs

```diff
@@ -0,0 +1,134 @@
+use crate::{
+    block::Block, db_state::SsTableId, filter::BloomFilter, flatbuffer_types::SsTableIndexOwned,
+};
+use async_trait::async_trait;
+use std::{sync::Arc, time::Duration};
+
+#[derive(Clone)]
+pub(crate) enum CachedBlock {
+    Block(Arc<Block>),
+    #[allow(dead_code)]
+    Index(Arc<SsTableIndexOwned>),
+    #[allow(dead_code)]
+    Filter(Arc<BloomFilter>),
+}
+
+#[derive(Clone, Copy)]
+pub enum CacheType {
+    Moka,
+}
+
+#[derive(Clone, Copy)]
+pub struct BlockCacheOptions {
+    pub max_capacity: u64,
+    pub cached_block_size: u32,
+    pub time_to_live: Option<Duration>,
+    pub time_to_idle: Option<Duration>,
+    pub cache_type: CacheType,
+}
+
+impl Default for BlockCacheOptions {
+    fn default() -> Self {
+        Self {
+            max_capacity: 64 * 1024 * 1024, // 64MB default max capacity
+            cached_block_size: 32,          // 32 bytes default
+            time_to_live: None,
+            time_to_idle: None,
+            cache_type: CacheType::Moka, // Default to Moka cache
+        }
+    }
+}
+
+#[async_trait]
+pub(crate) trait BlockCache: Send + Sync + 'static {
+    async fn get(&self, key: (SsTableId, usize)) -> CachedBlockOption;
```

Comment:
i guess using `Option<CachedBlock>` would make this type signature more clear about it's returning an `Option`.

`CachedBlockOption` has defined `block` `sst_index` and `bloom_filter` methods as accessors. i guess these accessors could be put in the `CachedBlock` struct, and something like `let block = opt.flat_map(|b| b.block());` on the caller sites.

Reply:
Felt like .block() is a cleaner way to do that instead of writing flat_maps everywhere

Reply:
how about this interface?

```
pub(crate) trait BlockCache: Send + Sync + 'static {
    async fn get_block(&self, key: (SsTableId, usize)) -> Option<Block>
   
    async fn get_bloomfilter(&self, key: (SsTableId, usize)) -> Option<bloomfilter>
```

or even:

```
  async fn get<T: CacheableBlockData>(&self, key: (SsTableId, usize)) -> Option<T>
```

i guess that the caller mostly knows what he/she want to get (`block`, `sst_index`, or `bloom_filter`), so the code could become even shorter on the caller sites, and less checks are needed (i've already asked for `bloom_filter`, and i do not need to check the result is `bloom_filter` or not again).